### ISSUE TITLE
Fail-free notifications

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/seastar#254]
There is code which needs to notify peers (unblock them) reliably from a noexcept block. Such notification must not fail. One example is unblocking peers blocked on memory from a free() path.

Notification boils down to calling promise::set_value(), which calls schedule(std::unique_ptr<task>). Currently schedule() can fail, because add_task() may need to allocate memory when the pending task queue needs to be resized. Such resizing may fail.

The risk of schedule() failing could be reduced by using a data structure with fragmented storage. That could still fail, in Scylla we do this under memory reclaim lock. See scylladb/scylla#2021.

A way of dealing with failures could be to have an emergency pre-allocated queue aside which would be used if regular push() fails. That queue would have to have a limited capacity, and could theoretically overflow.

We could make schedule() fail-free completely by adding a `task*` link to `class task` which would be used to chain it into emergency linked-list in case regular push fails. The downside though is increase in size of every task by 8 bytes.

**Comment by nyh** [comment]
There are additional serious problems with the fact that schedule() may fail, especially issue #84 which can cause the entire application to abort() if schedule() fails inside a then() call.

So I think the really issue should be about "schedule() must be noexcept" rather than any specific bug caused by the fact it is not like #84 or this issue. 

Like you, I am worried that replacing the vector with a, say, chunked_fifo, will only reduce the likelihood of this problem but not eliminate it completely. So I liked your last suggestion of an using intrusive linked list, same as what we did in the OSv scheduler's run queue, where the link is inside the thread object. I am not sure it even needs to be an "emergency linked list" - can't we just use these links as the only way to link the tasks in ready queues, not just in case of emergencies?

I think the additional task size overhead is acceptable - my view is that having miilions of unready tasks is not a good Seastar coding style. That could happen, for example, if someone uses parallel_for_each on a million item instead of do_for_each with a semaphore to limit parallelism (see discussion of this in the tutorial) - this is bad form because of the memory overhead of each task can be much greater than just the 8 bytes we want to add now. So I think it's fine that we have a lot of small tasks, but there is no reason to assume that most of these tasks will be unready - and if they are ready, they will have the queuing overhead in any case.

**Comment by nyh** [comment]
Hmm, looking again at the code, it seems to me the problem isn't just the queue of read continuations - I think schedule() also needs to allocate a copy of the continuation. How do we work around that?

**Comment by tgrabiec** [comment]
@nyh I think there are two different problems here, and they're not symmetric. There are two schedule() functions in this context, `promise::schedule()` and `::schedule(std::unique_ptr<task>)`. The one used by notification side (`promise::set_value()`) calls the latter, which calls `add_task()`. We could make this no-except as discussed earlier. The `task` object is allocated earlier in a different context, by `promise::schedule()`. That needs to allocate memory for the task and it's a much more difficult problem to make that fail-free.

In theory, for some uses, it would be fine if `promise::schedule()` fails, while it is not fine for `promise::set_value()` to fail. The fiber which tries to obtain a resource may fail to block, but notification that the resource was released cannot fail. It's similar as with object construction / destruction. One can fail, the other not.

One reason why it's not fine for `promise::schedule()` to fail in general is that we also use callbacks to install work which must be run later (there is not option to not run it, or run it now instead of later). One such use is to implement async scoping. For example, in a threading model you would write:

```
semaphore sem;
{
   auto units = sem.get_units(1);
   process();
}
```

In seastar you write:

```
units = sem.get_units(1);
return process().then([units = std::move()] {});
```

Here, then() must not fail, because we units must go out of scope only after process() completes, otherwise the program is not correct. At the point we discover we cannot fulfill this then(), there is no way to abort process().

This problem could be avoided by allocating resources up front, for example like this:

```
  auto units = sem.get_units(1)
  auto after_process = prepare([units = std::move(units)] {}); // allocates task<>, can fail
  return process()
      .then(std::move(after_process)); //cannot fail
```

That's equivalent (though more expensive) to a thread reserving space on stack before it enters a scope.

**Comment by avikivity** [comment]
Good observation and suggestion.  We could overload then() to be more careful with the result of prepare().

Hopefully we can hide most of the uses of this construct in new primitives.
